### ISSUE TITLE
user defined node

### ISSUE THREAD

**Issue by yakaibai** [Hopsan/hopsan#2128]
Is there a way that user can define a new node without the compiling of the whole package. I have worked like this:
```c++
//! @brief A energy node
//! @ingroup NodeEnergy
class NodeEnergy :public Node
{
public:
    //! @brief The data variable indexes, DataLength is used internally
    //! @ingroup NodeEnergy
    enum DataIndexEnumT {mFlow, Pressure, Temperature, DataLength};
    static Node* CreatorFunction() {return new NodeEnergy;}

private:
    NodeEnergy() : Node(DataLength)
    {
        setNiceName("energy");
        setDataCharacteristics(mFlow, "mFlow", "mdot", "mFlow", FlowType);
        setDataCharacteristics(Pressure, "Pressure", "p", "Pressure", IntensityType);
        setDataCharacteristics(Temperature, "Temperature", "T", "Temperature");
    }
};
```

  However, the whole program broke down when I pull this new component on the screen.

**Comment by peterNordin** [comment]
Have you implemented this new node inside a custom component library or directly inside the Hopsan source code?
The original intention was to be able to add nodes inside custom component libraries, but I am not sure if it still works.

If you have implemented it directly in the Hopsan code then you must recompile everything.

If you have implemented it inside a custom component, then HopsanGUI should be able to recompile that component library if you right click the library and choose recompile (in the library widget). But this requires that the source code files are listed in the library xml file.

Do you get any error messages when trying to connect ? Turn on debug output also (the blue/purple button in the message widget) ?
Can you make the source code available (on github or zip file)? Then I may be able to to find the issue.

What are you trying to connect to? You will need two components with ports using this new node type.
Possibly these components need to have read/write ports configured (if this is used as a signal node) or PowerPorts if you are attempting to implement C/Q components with the TLM method.

**Comment by peterNordin** [comment]
hmmm I tried to add your code to the example library and add two components, but the ports does not show up at all on the new components.
Is this the same problem that you face, that there are no visible ports to connect to ?

**Comment by peterNordin** [comment]
I think I found the problem. Hopsan did not find a matching port icon, and then no graphics item shows up.
I have a test branch, https://github.com/peterNordin/hopsan/commits/fix-add-custom-node

I dont know if you have already built the entire hopsan package, then can test that branch?
Otherwise you will have to wait until I have time to build a snapshot release.
What version are you using? build from source, Ubuntu or Windows ?

I also added a quick test of your code in commit: https://github.com/peterNordin/hopsan/commit/365a931dce7eea0def103d0d06d516043011fd97
But my other port graphics fix is needed for it to work.
